zrml/neo-swaps/src/tests/buy_and_sell.rs

```diff
@@ -0,0 +1,182 @@
+// Copyright 2023 Forecasting Technologies LTD.
+//
+// This file is part of Zeitgeist.
+//
+// Zeitgeist is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by the
+// Free Software Foundation, either version 3 of the License, or (at
+// your option) any later version.
+//
+// Zeitgeist is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Zeitgeist. If not, see <https://www.gnu.org/licenses/>.
+
+use super::*;
+use zeitgeist_primitives::constants::BASE;
+
+macro_rules! assert_pool_status {
+    ($market_id:expr, $reserves:expr, $spot_prices:expr, $fees:expr $(,)?) => {
+        let pool = Pools::<Runtime>::get($market_id).unwrap();
+        assert_eq!(pool.reserves.values().cloned().collect::<Vec<u128>>(), $reserves);
+        assert_eq!(
+            pool.assets()
+                .iter()
+                .map(|&a| pool.calculate_spot_price(a).unwrap())
+                .collect::<Vec<u128>>(),
+            $spot_prices,
+        );
+        let invariant = $spot_prices.iter().sum::<u128>();
+        assert_approx!(invariant, _1, 1);
+        assert_eq!(pool.liquidity_shares_manager.fees, $fees);
+    };
+}
+
+#[test]
+fn buy_and_sell() {
+    ExtBuilder::default().build().execute_with(|| {
+        let asset_count = 3;
+        let market_id = create_market_and_deploy_pool(
+            ALICE,
+            BASE_ASSET,
+            MarketType::Categorical(asset_count),
+            _100,
+            vec![_1_2, _1_4, _1_4],
+            CENT,
+        );
+        assert_ok!(AssetManager::deposit(BASE_ASSET, &ALICE, _1000));
+        assert_ok!(AssetManager::deposit(BASE_ASSET, &BOB, _1000));
+        assert_ok!(AssetManager::deposit(BASE_ASSET, &CHARLIE, _1000));
+
+        assert_ok!(NeoSwaps::buy(
+            RuntimeOrigin::signed(ALICE),
+            market_id,
+            asset_count,
+            Asset::CategoricalOutcome(market_id, 2),
+            _10,
+            0,
+        ));
+        assert_pool_status!(
+            market_id,
+            vec![598_000_000_000, 1_098_000_000_000, 767_092_556_931],
+            [4_364_837_956, 2_182_418_978, 3_452_743_066],
+            1_000_000_000,
+        );
```

Comment:
I think this macro makes the test hard to understand.

Reply:
Yeah, maybe. But it does eliminate a low of duplication _and_ ensures that we don't forget to check any of these values. Actually, I don't think it makes it harder to read. Replacing this macro invocation with the same six to eight lines over and over again won't improve readability.

I might also add a couple of tests at some point with a different number of assets. This macro will help there as well.